Adding example for Cloud Dataproc

evilsoapbox opened the pull request:
Go example for Google Cloud Dataproc. This sample follows a pretty basic (and common) Cloud Dataproc pattern:
1. Create cluster
2. Submit job
3. Get job output
4. Delete cluster

A sample PySpark job is also included in the repository.

broady commented:
High level comments:

1. Is there a better name for this sample? We're trying to avoid putting samples at the top level directory: it makes it more difficult to add more samples for the same product again in the future. See #75 for a recent example.
2. Please use the `cloud.google.com/go/storage` package for GCS.
3. Package level doc should be in the form of `Command foo ...` or `Sample foo ...`
4. Please use the license header found in all the other files.
5. Function-level comments should be in this form:
```
CreateCluster creates a Cloud Dataproc cluster with the given name, region, project and zone.
func CreateCluster(...
```
6. Thanks for writing a test, but the test provided doesn't actually test anything. See the tests in #75 for something more appropriate for this kind of sample.
7. Only the `main` function should call `log.Fatal`.
8. Please put the `main` function at the top of the file. It's difficult to recognize the program flow otherwise.
9. Error variables should be named `err` unless there's a good reason for them to be named something else.
10. I believe we use the term `region`, not `regionID` (across Cloud).

evilsoapbox commented:
1: How about putting it under something like `/data/dataproc/...` which would also allow other GCP data products, like Cloud Dataflow, to use the base directory `/data/dataflow/...` or BigQuery `/data/bigquery/...`. LMK what you think and I will reorganize.
2: Fixed
3: Fixed
4: Fixed
5: Fixed
6: Fixed
7: Fixed
8: Fixed
9: Fixed
10: Fixed

I just pushed an update to my fork based on the feedback.

As a note, in the unit tests the `TestJobFunctionality()` method is pretty weighty. The thinking here is all ops are dependent on one another and not well run atomically. Let me know if it's too weighty.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+// getDataprocService creates and returns a Cloud Dataproc service object.
+func getDataprocService() (service *dataproc.Service, err error) {
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, scope)
```
This error is ignored. I suggest inlining this (move all of this up to `main()`) too.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+// getStorageService creates and returns a Cloud Storage service object.
+func getStorageService() (service *storage.Client, err error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
```
Same here.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+
+	// Create a cluster
+	_, err = createCluster(service, *clusterName, *region, *projectID, *zoneID)
+	if err != nil {
```
`if _, err := createCluster(...); err != nil {`

evilsoapbox replied:
Good catch. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println("Cluster created")
```
log.Print?

evilsoapbox replied:
Yeah. That makes more sense. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("Job output:\n%s\n", string(output))
```
No need for `string()` - you're already using `%s` in the format string.

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	output, err := getJobOutput(storageService, *projectID, clusterData[1], clusterData[3], jobId)
```
what are 1 and 3 here?

evilsoapbox replied:
Yeah, I should be explicit. They are UUIDs for a cluster and bucket. I have added an explicit assignment so someone reading this knows what they are. (If this should be a Map, I can do that too.)
```
	var clusterUuid = clusterData[1]
	var bucketUuid = clusterData[3]
	if output, err := getJobOutput(storageService, *projectID, clusterUuid, bucketUuid, jobId); err != nil {
```

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+		log.Fatal(err)
+	}
+	fmt.Println("Cluster deleted")
+	fmt.Printf("Done...")
```
`Done..` is quite extraneous.

evilsoapbox replied:
Removed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+
+// createCluster creates a Cloud Dataproc cluster with the given name and region.
+func createCluster(service *dataproc.Service, name string, region string,
+	project string, zone string) (response string, err error) {
```
Wrapping lines is a good sign that the parameters should go into a struct. These get passed around an awful lot.

Would something like this make sense?

```
type clusterConfig struct{
    project, region, zone, name string
}
```

evilsoapbox replied:
Implemented throughout.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	}
+
+	// Find the cluster requested in the list
+	for _, item := range clusters {
```
`for _, c := range clusters`

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+		}
+	}
+
+	return nil, errors.New("Cluster not found")
```
Error strings should start with a lowercase letter. They are often concatenated with a prefix - having a capital in the middle is ugly.

evilsoapbox replied:
Ack. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	object := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", cluster, job)
+
+	// Read the file
+	rc, err := storageClient.Bucket(bucket).Object(object).NewReader(context.Background())
```
`r, err := `

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+func getJobOutput(storageClient *storage.Client, project string, cluster string, bucket string, job string) (output []byte, err error) {
+
+	// Format the object name
+	object := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", cluster, job)
```
Would be good to have a reference link here as a comment - where does this string format come from?

evilsoapbox replied:
Ack. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	// List all clusters in a project for a given region
+	res, err := service.Projects.Regions.Clusters.List(*projectID, *region).Do()
+
+	for _, item := range res.Clusters {
```
`for _, c := range ...`

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service) (clusters [][]string, err error) {
+	// List all clusters in a project for a given region
+	res, err := service.Projects.Regions.Clusters.List(*projectID, *region).Do()
```
Check this error.

broady replied:
Also inconsistent use of flags vs function parameters (`*service, *region`)

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	//Get the Job's status
+	res, err := service.Projects.Regions.Jobs.Get(project, region, jobId).Do()
+
+	return res.Status.State, err
```
I'm sure this will result in a nil pointer panic.

Check the error.

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+
+// getJobStatus returns the status of the job with the given job id, project, and region.
+func getJobStatus(service *dataproc.Service, jobId string, project string, region string) (status string, err error) {
+	//Get the Job's status
```
Add a space after `//`

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	wc.ContentType = "text/plain"
+	wc.ACL = []storage.ACLRule{{storage.AllUsers, storage.RoleReader}}
+	file, err := ioutil.ReadFile(filepath)
+	if err != nil {
```
Read the file first - then you don't have the racy `os.Stat` call

evilsoapbox replied:
Ack. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+}
+
+// submitJob submits a PySpark job with the given file path to a PySpark file, project, bucket, and cluster.
+func submitJob(service *dataproc.Service, storageClient *storage.Client, filepath string, project string, bucket string, cluster string) (jobId string, err error) {
```
`jobId` -> `jobID`

evilsoapbox replied:
Good catch. Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+		Job: &job,
+	}
+
+	_, err = service.Projects.Regions.Jobs.Submit(*projectID, "global", &jobRequest).Do()
```
Use `project` from the parameters.

evilsoapbox replied:
Fixed with struct.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+	pySparkJob := dataproc.PySparkJob{
+		MainPythonFileUri: "gs://" + bucket + "/" + filename,
+	}
+	jobID := "test-job-" + fmt.Sprintf("%v", time.Now().Unix())
```
Why's it a `test-job`?

evilsoapbox replied:
Fair point. I removed the prefix so the job name is the unix time.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+		}
+
+		// Sleep for one second
+		time.Sleep(1000 * time.Millisecond)
```
`time.Second`, remove comment.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+			return false, err
+		}
+		if jobStatus == "DONE" {
+			finished = true
```
return true, nil

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+			finished = true
+		} else if jobStatus == "ERROR" {
+			finished = true
+			err = errors.New("Job finished with an error")
```
return true, errors.New("job errored")

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,313 @@
+
+// waitForJob waits for a job to finish with the given job id, project, and region.
+func waitForJob(service *dataproc.Service, jobId string, project string, region string) (finished bool, err error) {
+	for finished == false {
```
just
```
for {
   // ...
}
```

(use early returns)

evilsoapbox replied:
Sweet, done.

broady commented:
@evilsoapbox thanks!

Does `golang-samples/dataflow/runjob` make sense? I like the idea of keeping bigquery top-level, so dataflow probably should be, too.

I'll look at the tests again later - probably tomorrow.

evilsoapbox commented:
@broady  Thanks for taking a look!

Just added a new commit with fixes and revised tests.

How about `golang-samples/dataproc/runjob`? Dataflow is a seperate GCP product and they may want to use that name. :)

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	scope = "https://www.googleapis.com/auth/cloud-platform"
+
+	// Compute URI base
+	computeUriBase = "https://www.googleapis.com/compute/v1/projects/%s/zones/%s"
```
`s/Uri/URI`

Also, it's not really a base. It's a format string.

`computeURIFormat = ...`?

evilsoapbox replied:
Good point. Added some clarification and a link to the docs.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+		name,
+		uuid,
+		zone string
+}
```
Run `gofmt -s .` from the root directory.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+		log.Fatal(err)
+	}
+
+	// Create a new Google Clout Storage service
```
ouch!

Either `s/Clout/Cloud`, or better: remove these comments. It's clear that it's creating a new X service.

evilsoapbox replied:
Done!

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// Create a new Google Clout Storage service
+	storageService, err := storage.NewClient(ctx)
```
`storageClient`

evilsoapbox replied:
Done

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	cluster := clusterConfig{project: *projectID, region: *region, name: *clusterName, zone: *zoneID}
+
+	// Create a cluster
+	if _, err := createCluster(service, &cluster); err != nil {
```
Add ctx as a first parameter to all of these.

broady replied:
On all of these. No need to take the pointer to the cluster variable. Just pass the struct through.

evilsoapbox replied:
Updated to point to the struct without pointers.

With respect to the `ctx` to you mean line 86 in the `createCluster` and other methods? I'm new to Go and am looking for a design pattern around it but can't find a great example so I am a bit stuck there. :)

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	log.Println("Cluster created")
+
+	// Wait on the cluster to become active
+	log.Println("Waiting for cluster to be ready")
```
Again, the comment says the same thing as the log statement. Remove the comments.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	log.Printf("Job output:\n%s\n", output)
+
+	// Delete the cluster
+	if _, err = deleteCluster(service, &cluster); err != nil {
```
Change `err =` to `err :=` (reduces scope)

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	// Create the cluster
+	res, err := service.Projects.Regions.Clusters.Create(cluster.project, cluster.region, &clusterSpec).Do()
+
+	return fmt.Sprintf("%s", res), err
```
Just `return ...Do()`, perhaps? The formatted string is not used except for tests.

evilsoapbox replied:
Removed the string output and just return the error now.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+
+// getJobOutput returns the text from the job (raw driver output) with the given project, cluser name, bucket id, and job id.
+func getJobOutput(storageClient *storage.Client, job string, cluster *clusterConfig) (output []byte, err error) {
+
```
Remove blank line.

evilsoapbox replied:
Done.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+
+// deleteCluster deletes the Cloud Dataproc cluster with the given project, region, and name.
+func deleteCluster(service *dataproc.Service, cluster *clusterConfig) (response string, err error) {
+	// Delete the cluster
```
Remove this comment

evilsoapbox replied:
Done

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service, project string, region string) (clusters [][]string, err error) {
```
A struct would be nicer than a `[][]string`

evilsoapbox replied:
Refactored to introduce a second struct `clusterDetails` to contain details about a created cluster. This cleanly (I think) separates a `config` which is used to create a cluster with the actual returned details from a cluster.

Unfortuinately, our API does not cleanly return `region` or `zone` so I have to parse some URLs to get them.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// Upload the file to GCS
+	job_filename_parts := strings.Split(filepath, "/")
```
Camel case, please

evilsoapbox replied:
Fixed.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+// updateClusterMetadata geta metadata and updates the given clusterConfig.
+func updateClusterMetadata(service *dataproc.Service, cluster *clusterConfig) (err error) {
```
Ah, yes, this one will need a pointer.

evilsoapbox replied:
Fixed with the new struct.

broady commented on dataproc/dataproc.go:
```diff
@@ -0,0 +1,304 @@
+		state,
+		name,
+		uuid,
+		zone string
```
A few of these fields are only set by the server, like uuid. I think only the user-set fields should be in a struct named `XConfig`

evilsoapbox replied:
Fixed with the new struct.

broady commented:
This is looking very good. Thank you.

evilsoapbox commented:
Thank you @broady for all of the feedback (and sorry for being a pain - coming to go from Python.) :) Just pushed a new commit with fixes.

broady commented:
Thanks, @evilsoapbox. No worries! I'll probably take over the CL and do the final clean up and merge.

Gimme a few days at least. A bit backed up right now.

shimont commented:
This example really helps!
Why dont you merge it?

odeke-em commented:
How's it going @broady? Just a gentle ping here.

tbpg commented:
Hi Folks. Sorry this PR fell off of the radar. Would you like to move forward with it?

I'm going to close it for now (it's been ~2 years since the last update). But, if we still need/want it and it will be embedded in documentation somewhere, we can reopen, clean it up, and move forward.

cc @bradmiro
